File: 859.buddy-strings.js
/*
 * @lc app=leetcode id=859 lang=javascript
 *
 * [859] Buddy Strings
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} goal
 * @return {boolean}
 */
var buddyStrings = function(s, goal) {
  if (s.length !== goal.length) return false
  if (s === goal) {
    s = [...s].sort()
    for (let i=1; i<s.length; i++) {
      if (s[i-1] === s[i]) return true
    }
    return false
  }

  let badIndexes = []
  for (let i=0; i<s.length; i++) {
    if (s[i] !== goal[i]) {
      badIndexes.push(i)
      if (badIndexes.length > 2) return false
    }
  }

  if (badIndexes.length !== 2) return false

  let [i1, i2] = badIndexes
  let a = s[i1]
  let b = s[i2]
  s = s.slice(0, i1) + b + s.slice(i1+1, i2) + a + s.slice(i2+1)
  return s === goal
};
// @lc code=end

File: 859.buddy-strings.py
#
# @lc app=leetcode id=859 lang=python3
#
# [859] Buddy Strings
#

# @lc code=start
class Solution:
    def buddyStrings(self, s: str, goal: str) -> bool:
        if len(s) != len(goal):
            return False
        if s == goal:
            ss = sorted(s)
            for i in range(1,len(ss)):
                if ss[i-1] == ss[i]:
                    return True
            return False

        badIndexes = []
        for i,c in enumerate(s):
            if (c != goal[i]):
                badIndexes.append(i)
            if len(badIndexes) > 2:
                return False

        if len(badIndexes) < 2:
            return False

        i,i2 = badIndexes
        s2 = s[:i] + s[i2] + s[i+1:i2] + s[i] + s[i2+1:]

        return s2 == goal
# @lc code=end

